### PR TITLE
enable both hints and default values in shader language

### LINKED ISSUE #15931: Shader Language: Cannot use hints and default values at the same time
**Godot version:**
3.0.beta.custom_build.a60898 (should be true for all 3.0 builds)

**OS/device including version:**
Windows 10, 64-bit

**Issue description:**
When writing a shader, it is possible to specify hints for uniforms:
```uniform vec4 color : hint_color;```

In the shader params section of the ShaderMaterial, the control to modify the value is then a color picker. However, in many cases, we may also want to specify default values. Usually, this is possible using the following syntax:

```uniform vec4 color = vec4(1.0,0.0,1.0,1.1);```

However, there appears to be no possibility to combine the two approaches, neither of the two statements work:
```uniform vec4 color : hint_color =  vec4(1.0,0.0,1.0,1.0);```
```uniform vec4 color = vec4(1.0,0.0,1.0,1.0) : hint_color;```

Maybe I missed something, if not, that'd be a really helpful feature, in my opinion.

**Steps to reproduce:**
* Create a MeshInstance
* Assign a new ShaderMaterial
* Create a new Shader:
```
shader_type spatial;

render_mode unshaded;

uniform vec4 color : hint_color =  vec4(1.0,0.0,1.0,1.1);

void fragment() {
	ALBEDO = color.rgb;
}
```
* Realize it will not compile because it expects a semicolon after the hint.

**Minimal reproduction project:**
[shader_hint_default_value.zip](https://github.com/godotengine/godot/files/1649678/shader_hint_default_value.zip)

### PR DESCRIPTION
**Pull request by brainsick**
Needs review.  Discussion in attached issue.

Fixes #15931.

Best reviewed [without](https://github.com/godotengine/godot/pull/16066/files?w=1) whitespace.

### CODE REVIEW

File: servers/visual/shader_language.cpp [hunk 1]
```diff
@@ -3661,139 +3661,148 @@ Error ShaderLanguage::_parse_shader(const Map<StringName, FunctionInfo> &p_funct
...
-
-						uniform.default_value.resize(cn->values.size());
-
-						if (!convert_constant(cn, uniform.type, uniform.default_value.ptrw())) {
-							_set_error("Can't convert constant to " + get_datatype_name(uniform.type));
-							return ERR_PARSE_ERROR;
-						}
-						tk = _get_token();
-					}
+					while (true) {
```
Comment (groud) [inline servers/visual/shader_language.cpp]: Why not just write `while (tk.type == TK_OP_ASSIGN)` ?

Comment (brainsick) [reply servers/visual/shader_language.cpp]: No compelling reason, honestly.  I was just trying to follow existing conventions and `while (true)` was prevalent.  If that's the only issue, I can definitely address that.

Comment (groud) [reply servers/visual/shader_language.cpp]: Well, it depends. But here it make the code a lot more longer and far less readable. So yeah, I would prefer replacing by `while (tk.type == TK_OP_ASSIGN)`. :)
Please amend your commit (no need to create a new one)

Comment (brainsick) [reply servers/visual/shader_language.cpp]: Wait, it doesn't work.  The first time through the loop is the colon operator.  The second time through is the assignment.

Comment (groud) [reply servers/visual/shader_language.cpp]: Arf sorry. My bad, I did not unfolded the code XD.

I would personally have put the `break` inside each `if`, or a least get rid of the case with the `continue` as it makes no sense at the end of the loop. But, anyway, that's ok like that, sorry for the inconvenience.

**Comment by hpvb** [comment]
We merged #17533 instead. Thanks for your contribution though!

**Pull request closed without merging**